src/Workspaces/Core/Portable/Storage/SQLite/v2/Interop/SqlStatement.cs

```diff
@@ -67,7 +69,39 @@ public Result Step(bool throwOnError = true)
         }
 
         internal void BindStringParameter(int parameterIndex, string value)
-            => _connection.ThrowIfNotOk(NativeMethods.sqlite3_bind_text(_rawStatement, parameterIndex, value));
+        {
+            // Attempt to stackalloc utf8 converted small strings to avoid lots of allocs.
+            //
+            // This is safe as sqlite will still copy our bytes over to its own internal memory (see
+            // conversation here: https://github.com/dotnet/roslyn/pull/51111#pullrequestreview-588169715)
+            // on the topic.  So it's fine that our own stackalloc'ed bytes will be gone when this function
+            // returns and our caller continues to interact with this SqlStatement instance.
+            var utf8ByteCount = Encoding.UTF8.GetByteCount(value);
+            var utf8WithTrailingZeroByteCount = utf8ByteCount + 1;
+
+            if (utf8WithTrailingZeroByteCount <= 512)
+            {
+                Span<byte> bytes = stackalloc byte[utf8WithTrailingZeroByteCount];
+#if NETCOREAPP
+                Contract.ThrowIfFalse(Encoding.UTF8.GetBytes(value.AsSpan(), bytes) == utf8ByteCount);
+#else
+                unsafe
+                {
+                    fixed (char* charsPtr = value)
+                    fixed (byte* bytesPtr = bytes)
+                    {
+                        Contract.ThrowIfFalse(Encoding.UTF8.GetBytes(charsPtr, value.Length, bytesPtr, utf8WithTrailingZeroByteCount) == utf8ByteCount);
+                    }
+                }
+#endif
+                bytes[utf8WithTrailingZeroByteCount - 1] = 0;
```

Comment:
There is no need for a trailing null byte. Per [the code](https://github.com/ericsink/SQLitePCL.raw/blob/ced6c13984f604f6d91d27f1c0f46a06ada46e37/src/SQLitePCLRaw.provider.dynamic_cdecl/Generated/provider_dynamic_cdecl.cs#L1181) the length comes from the length of the span passed in. In fact as it is now, you may be inadvertently *including* the null terminator in your stored string.

Reply:
hrmm... definitely confusing as utf8z requires a trailing null.  But your reading of things seems correct to me.  I'll tweak.

Reply:
I tested it [here](https://dev.azure.com/devdiv/DevDiv/_git/VS.CloudCache/pullrequest/303830) and it worked. I even put extra non-null bytes after the span passed into the method and they didn't show up, so it definitely truncates with the length rather than a null.